### PR TITLE
Fix udf-compiler scala2.13 internal return statements

### PR DESCRIPTION
**Pull request by abellina**
Closes https://github.com/NVIDIA/spark-rapids/issues/11174

This is in draft to solicit comments from @seanprime7

### CODE REVIEW

File: udf-compiler/src/main/scala/com/nvidia/spark/udf/CFG.scala [hunk 1]
```diff
@@ -171,72 +194,88 @@ object CFG {
       val nextOffset: Int = codeIterator.lookAhead
       val opcode: Int = codeIterator.byteAt(offset)
       // here we are looking for branching instructions
-      opcode match {
-        case Opcode.IF_ICMPEQ | Opcode.IF_ICMPNE | Opcode.IF_ICMPLT |
-             Opcode.IF_ICMPGE | Opcode.IF_ICMPGT | Opcode.IF_ICMPLE |
-             Opcode.IFEQ | Opcode.IFNE | Opcode.IFLT | Opcode.IFGE |
-             Opcode.IFGT | Opcode.IFLE | Opcode.IFNULL | Opcode.IFNONNULL =>
-          // an if statement has two other offsets, false and true branches.
+      if ((opcode == Opcode.ARETURN ||
```
Comment (seanprime7) [inline udf-compiler/src/main/scala/com/nvidia/spark/udf/CFG.scala:197]: This could be just added to the `opcode match {` below as a `case`?
That would minimize diff in the PR

**Comment by abellina** [comment]
@seanprime7 I made the suggested change, please take another look when you get a chance.

File: scala2.13/udf-compiler/pom.xml [hunk 2]
```diff
@@ -130,9 +130,6 @@
             <plugin>
                 <groupId>org.apache.maven.plugins</groupId>
                 <artifactId>maven-surefire-plugin</artifactId>
-                <configuration>
-                    <skipTests>true</skipTests>
-                </configuration>
```
Comment (jlowe) [inline scala2.13/udf-compiler/pom.xml:135]: Scala 2.13 pom was edited directly rather than generated, causing all the 2.13 builds to fail.

I'm not sure why this change was made, since the comment just above it seems to still apply:
```
            <!-- disable surefire as we are using scalatest only -->
```

File: udf-compiler/src/main/scala/com/nvidia/spark/udf/CFG.scala
Comment (jlowe) [inline udf-compiler/src/main/scala/com/nvidia/spark/udf/CFG.scala:1]: 2024 copyrights

File: udf-compiler/src/test/scala/com/nvidia/spark/OpcodeSuite.scala
Comment (jlowe) [inline udf-compiler/src/test/scala/com/nvidia/spark/OpcodeSuite.scala:1]: 2024 copyrights

File: scala2.13/udf-compiler/pom.xml (hunk 2 above)
Comment (abellina) [reply scala2.13/udf-compiler/pom.xml]: Updated here: https://github.com/NVIDIA/spark-rapids/pull/11553/commits/d26258b2299100e2ed78157ac420b4f84ae213ed

**Comment by abellina** [comment]
build

File: udf-compiler/src/test/scala/com/nvidia/spark/OpcodeSuite.scala [hunk 3]
```diff
@@ -36,6 +36,7 @@ class OpcodeSuite extends AnyFunSuite {
   val conf: SparkConf = new SparkConf()
     .set("spark.sql.extensions", "com.nvidia.spark.udf.Plugin")
     .set("spark.rapids.sql.udfCompiler.enabled", "true")
+    .set("spark.sql.ansi.enabled", "false")
```
Comment (jlowe) [inline udf-compiler/src/test/scala/com/nvidia/spark/OpcodeSuite.scala:39]: Curious why we need to disable ANSI here?  I assume this is for Spark 4.0, but then I would expect there to be corresponding negative tests where we are testing that we do _not_ replace UDFs because of ANSI expression semantics.  Seems like we need a comment and/or pointer to a tracking issue here.

Comment (abellina) [reply udf-compiler/src/test/scala/com/nvidia/spark/OpcodeSuite.scala:39]: correct that is due to spark 4, let me file something at least with the failures I see with ansi enabled.

File: udf-compiler/src/main/scala/com/nvidia/spark/udf/CFG.scala [hunk 4]
```diff
@@ -327,4 +366,13 @@ object CFG {
         successor + (src -> dst))
     }
   }
+
+  private def isReturn(opcode: Int): Boolean = {
```
Comment (seanprime7) [inline udf-compiler/src/main/scala/com/nvidia/spark/udf/CFG.scala:370]: I think you can use https://www.javassist.org/html/javassist/bytecode/analysis/Util.html#isReturn(int).
Can you check if it does the same thing you are trying to do here?

Comment (seanprime7) [reply udf-compiler/src/main/scala/com/nvidia/spark/udf/CFG.scala:370]: it seems to be the case based on https://github.com/jboss-javassist/javassist/blob/master/src/main/javassist/bytecode/analysis/Util.java#L45-L47

File: udf-compiler/src/test/scala/com/nvidia/spark/OpcodeSuite.scala (hunk 3 above)
Comment (abellina) [reply udf-compiler/src/test/scala/com/nvidia/spark/OpcodeSuite.scala:39]: https://github.com/NVIDIA/spark-rapids/issues/11633

**Comment by abellina** [comment]
build

**Comment by abellina** [comment]
build

**Comment by abellina** [comment]
This should be ready for another look @jlowe @seanprime7 when you get a chance.

**Review by jlowe — APPROVED** [review]

**Review by revans2 — APPROVED** [review]
